File: frontend/src/lib/mutations/meal.ts
import { mutate } from '$lib/mutate';
import { db } from '$lib/db/database';
import { todayString, nowIso } from '$lib/utils/datetime';
import { startOfLocalDayMs, userTimezone } from '$lib/utils/timezone';
import { SELF_USER_ID } from '$lib/constants';
import { uuid7 } from '$lib/db/uuid';

const MEAL_SOURCE = 'meal';

export interface MealItemInput {
  food_item_id: string;
  servings: number;
  amount_g?: number;
}

export async function createMeal(data: {
  log_date?: string;
  meal_type?: string;
  name?: string;
  notes?: string;
  items: MealItemInput[];
}) {
  const id = uuid7();
  const now = nowIso();
  const logDate = data.log_date ?? todayString();
  const measurementId = uuid7();
  const items = (data.items ?? []).map((item) => ({ id: uuid7(), ...item }));

  const macros = await calcMacros(items);
  const startTime = new Date(startOfLocalDayMs(logDate, userTimezone())).toISOString();

  return mutate({
    kind: 'command',
    command: 'create_meal',
    queueable: true,
    payload: {
      id,
      log_date: logDate,
      meal_type: data.meal_type ?? 'snack',
      name: data.name ?? null,
      notes: data.notes ?? null,
      measurement_id: measurementId,
      items: items.map((i) => ({
        id: i.id,
        food_item_id: i.food_item_id,
        servings: i.servings,
        amount_g: i.amount_g ?? null
      }))
    },
    optimisticTable: 'meal',
    optimisticData: {
      id,
      user_id: SELF_USER_ID,
      log_date: logDate,
      meal_type: data.meal_type ?? 'snack',
      name: data.name ?? null,
      notes: data.notes ?? null,
      created_at: now,
      updated_at: null,
      deleted_at: null
    },
    optimisticRows: [
      {
        table: 'meal_item',
        rows: items.map((item) => ({
          id: item.id,
          meal_id: id,
          user_id: SELF_USER_ID,
          food_item_id: item.food_item_id,
          servings: item.servings,
          amount_g: item.amount_g ?? null,
          created_at: now,
          deleted_at: null
        }))
      },
      {
        table: 'measurement',
        rows: [
          {
            id: measurementId,
            user_id: SELF_USER_ID,
            metric_code: 'nutrition',
            source_data_type: 'nutrition',
            source: MEAL_SOURCE,
            value_numeric: null,
            value_text: null,
            value_json: JSON.stringify(macros),
            start_time: startTime,
            end_time: null,
            notes: null,
            external_id: id,
            created_at: now,
            updated_at: null,
            deleted_at: null
          }
        ]
      }
    ],
    responseTable: 'meal'
  });
}

export async function updateMeal(
  mealId: string,
  data: {
    meal_type?: string;
    name?: string;
    notes?: string;
    items?: MealItemInput[];
  }
) {
  return mutate({
    kind: 'command',
    command: 'update_meal',
    queueable: true,
    payload: { id: mealId, ...data },
    optimisticTable: 'meal',
    optimisticData: { id: mealId, ...data },
    responseTable: 'meal'
  });
}

export async function deleteMeal(mealId: string) {
  const items = await db.meal_item.where('meal_id').equals(mealId).toArray();
  const itemIds = items.map((i) => i.id);
  const measurements = await db.measurement.where('external_id').equals(mealId).toArray();

  return mutate({
    kind: 'command',
    command: 'delete_meal',
    queueable: true,
    payload: { id: mealId },
    optimisticTable: 'meal',
    optimisticData: { id: mealId, deleted_at: nowIso() },
    optimisticDelete: [
      ...(itemIds.length > 0 ? [{ table: 'meal_item', ids: itemIds }] : []),
      ...(measurements.length > 0
        ? [{ table: 'measurement', ids: measurements.map((m) => m.id) }]
        : [])
    ]
  });
}

async function calcMacros(
  items: { food_item_id: string; servings: number }[]
): Promise<{ calories: number; protein_g: number; carbs_g: number; fat_g: number }> {
  const foodIds = [...new Set(items.map((i) => i.food_item_id))];
  const foods = await db.food_item.bulkGet(foodIds);
  const foodMap = new Map(foods.filter(Boolean).map((f) => [f!.id, f!]));

  const total = { calories: 0, protein_g: 0, carbs_g: 0, fat_g: 0 };
  for (const item of items) {
    const food = foodMap.get(item.food_item_id);
    if (!food) continue;
    const factor = item.servings;
    total.calories += food.calories_per_serving * factor;
    total.protein_g += food.protein_g * factor;
    total.carbs_g += food.carbs_g * factor;
    total.fat_g += food.fat_g * factor;
  }
  return total;
}

File: src/salus/services/commands/meal.py
from __future__ import annotations

import json
from datetime import date
from typing import Any, TYPE_CHECKING

from salus.models.food import Meal, MealItem
from salus.models.measurement import Measurement
from salus.schemas.commands import CreateMealPayload, DeleteMealPayload, UpdateMealPayload
from salus.services._helpers import uid
from salus.services.command_registry import CommandResult, register
from salus.services.serialization import serialize_record
from salus.services.timezone import start_of_local_day, tz_for, user_today
from salus.utils import uuid7_str

if TYPE_CHECKING:
    from salus.repositories.unit_of_work import IUnitOfWork
    from salus.models.user import User

MEAL_SOURCE = "meal"
NUTRITION_METRIC = "nutrition"
NUTRITION_SOURCE = "nutrition"

_MEAL_FIELDS = (
    "id", "user_id", "log_date", "meal_type", "name", "notes",
    "created_at", "updated_at", "deleted_at",
)


def _calc_macros(items: list[MealItem], food_map: dict) -> dict:
    total = {"calories": 0.0, "protein_grams": 0.0, "carbs_grams": 0.0, "fat_grams": 0.0}
    for item in items:
        food = food_map.get(item.food_item_id)
        if food is None:
            continue
        factor = item.servings
        total["calories"] += food.calories_per_serving * factor
        total["protein_grams"] += food.protein_g * factor
        total["carbs_grams"] += food.carbs_g * factor
        total["fat_grams"] += food.fat_g * factor
    return total


def _resolve_food_items(uow: IUnitOfWork, items: list[MealItem]) -> dict:
    food_ids = {item.food_item_id for item in items}
    result: dict = {}
    for fid in food_ids:
        food = uow.food_items.get_by_id(fid)
        if food:
            result[fid] = food
    return result


def _measurement_for_meal(
    meal: Meal, items: list[MealItem], food_map: dict, tz, measurement_id: str | None
) -> Measurement:
    macros = _calc_macros(items, food_map)
    return Measurement(
        id=measurement_id,
        user_id=meal.user_id,
        metric_code=NUTRITION_METRIC,
        source_data_type=NUTRITION_SOURCE,
        source=MEAL_SOURCE,
        external_id=meal.id,
        value_json=json.dumps(macros),
        start_time=start_of_local_day(meal.log_date, tz),
    )


def _delete_measurement(uow: IUnitOfWork, meal_id: str) -> None:
    existing = uow.measurements.find_by_external_id(meal_id, MEAL_SOURCE)
    if existing:
        uow.measurements.delete(existing)


@register("create_meal")
class CreateMealHandler:
    def execute(self, uow: IUnitOfWork, user: User, payload: dict[str, Any]) -> CommandResult:
        data = CreateMealPayload.model_validate(payload)
        user_id = uid(user)
        tz = tz_for(uow.session, user_id)
        if not data.items:
            return CommandResult(status="error", message="Meal must have at least one item")

        log_date = date.fromisoformat(data.log_date) if data.log_date else user_today(uow.session, user_id)
        meal = Meal(
            id=data.id or uuid7_str(),
            user_id=user_id,
            log_date=log_date,
            meal_type=data.meal_type,
            name=data.name,
            notes=data.notes,
        )
        uow.meals.add(meal)

        items = []
        for item_data in data.items:
            item = MealItem(
                id=item_data.id or uuid7_str(),
                meal_id=meal.id or "",
                user_id=user_id,
                food_item_id=item_data.food_item_id,
                servings=item_data.servings,
                amount_g=item_data.amount_g,
            )
            uow.meal_items.add(item)
            items.append(item)

        food_map = _resolve_food_items(uow, items)
        uow.measurements.add(_measurement_for_meal(meal, items, food_map, tz, data.measurement_id))

        uow.commit()
        uow.session.refresh(meal)
        return CommandResult(
            status="created",
            id=meal.id,
            record=serialize_record(meal, list(_MEAL_FIELDS)),
        )


@register("update_meal")
class UpdateMealHandler:
    def execute(self, uow: IUnitOfWork, user: User, payload: dict[str, Any]) -> CommandResult:
        data = UpdateMealPayload.model_validate(payload)
        user_id = uid(user)
        meal = uow.meals.get_by_id(data.id)
        if not meal or meal.user_id != user_id:
            return CommandResult(status="not_found", message="Meal not found")

        if data.meal_type is not None:
            meal.meal_type = data.meal_type
        if data.name is not None:
            meal.name = data.name
        if data.notes is not None:
            meal.notes = data.notes
        uow.meals.add(meal)

        if data.items is not None:
            for old in uow.meal_items.find_by_meal(data.id):
                uow.meal_items.delete(old)

            items = []
            for item_data in data.items:
                item = MealItem(
                    id=item_data.id or uuid7_str(),
                    meal_id=data.id,
                    user_id=user_id,
                    food_item_id=item_data.food_item_id,
                    servings=item_data.servings,
                    amount_g=item_data.amount_g,
                )
                uow.meal_items.add(item)
                items.append(item)
        else:
            items = uow.meal_items.find_by_meal(data.id)

        _delete_measurement(uow, data.id)
        food_map = _resolve_food_items(uow, items)
        uow.measurements.add(
            _measurement_for_meal(meal, items, food_map, tz_for(uow.session, user_id), data.measurement_id)
        )

        uow.commit()
        uow.session.refresh(meal)
        return CommandResult(
            status="updated",
            id=meal.id,
            record=serialize_record(meal, list(_MEAL_FIELDS)),
        )


@register("delete_meal")
class DeleteMealHandler:
    def execute(self, uow: IUnitOfWork, user: User, payload: dict[str, Any]) -> CommandResult:
        data = DeleteMealPayload.model_validate(payload)
        meal = uow.meals.get_by_id(data.id)
        if not meal:
            return CommandResult(status="deleted", id=data.id)
        if meal.user_id != uid(user):
            return CommandResult(status="forbidden", message="Not your meal")

        for item in uow.meal_items.find_by_meal(data.id):
            uow.meal_items.delete(item)
        _delete_measurement(uow, data.id)
        uow.meals.delete(meal)
        uow.commit()
        return CommandResult(status="deleted", id=data.id)
